### PR TITLE
chore: migrate `tools` to TypeScript

### PR DESCRIPTION
**Pull request by Esemesek**
Summary:
---------

- [x] - setup Typescript build process
- [x] - migrate `@react-native-community/cli-tools` to Typescript

Test Plan:
----------

Everything builds and tests are passing.

### CODE REVIEW

File: packages/tools/src/groupFilesByType.ts [hunk 1]
```diff
@@ -4,22 +4,11 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  *
- * @flow
  */
 
-import {groupBy} from 'lodash';
+import { groupBy } from 'lodash';
```
Comment (thymikee) [inline packages/tools/src/groupFilesByType.ts]: fix spacing

File: packages/tools/src/groupFilesByType.ts [hunk 2]
```diff
@@ -4,22 +4,11 @@
...
 import mime from 'mime';
 
-/**
- * Since there are no officially registered MIME types
- * for ttf/otf yet http://www.iana.org/assignments/media-types/media-types.xhtml,
- * we define two non-standard ones for the sake of parsing
- */
-mime.define({
-  'font/opentype': ['otf'],
-  'font/truetype': ['ttf'],
```
Comment (thymikee) [inline packages/tools/src/groupFilesByType.ts]: why is this removed?

File: scripts/buildTs.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+const {getPackages, adjustToTerminalWidth, OK} = require('./helpers');
+
+const packages = getPackages();
+
+const packagesWithTs = packages.filter(p =>
+  fs.existsSync(path.resolve(p, 'tsconfig.json')),
+);
+
+const args = [
+  '--max-old-space-size=4096',
```
Comment (thymikee) [inline scripts/buildTs.js]: I think we have too small code base for this?

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    /* Module Resolution Options */
+    "moduleResolution": "node",
+    "esModuleInterop": true,
+    "resolveJsonModule": true
+  },
+  "exclude": [
+    ".yarn/releases/*",
+    "**/__tests__/**/*",
+    "**/build/**/*",
+    "**/build-es5/**/*"
```
Comment (thymikee) [inline tsconfig.json]: remove this

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    "noImplicitReturns": true,
+    "noFallthroughCasesInSwitch": true,
+
+    /* Module Resolution Options */
+    "moduleResolution": "node",
+    "esModuleInterop": true,
+    "resolveJsonModule": true
+  },
+  "exclude": [
+    ".yarn/releases/*",
```
Comment (thymikee) [inline tsconfig.json]: remove this

File: packages/tools/src/groupFilesByType.ts (hunk 2 above)
Comment (Esemesek) [reply packages/tools/src/groupFilesByType.ts]: I've updated version of `mimetype` which has those types buildin.

File: packages/tools/src/groupFilesByType.ts (hunk 1 above)
Comment (Esemesek) [reply packages/tools/src/groupFilesByType.ts]: I will need to configure linter with prettier for Typescript.

**Comment by grabbou** [comment]
Awesome! A small step for a human, big step for CLI xD

**Comment by lucasbento** [comment]
After this PR lands, can we start migrating the rest of the stuff away from Flow?

**Comment by grabbou** [comment]
Yes, that's the idea!

File: .eslintrc.js [hunk 6]
```diff
@@ -16,6 +16,8 @@ module.exports = {
       alias: {
         map: [['types', './types/index.js']],
       },
+      // Use <rootDir>/tsconfig.json for typescript resolution
+      typescript: {},
```
Comment (thymikee) [inline .eslintrc.js]: we should add TS plugins for ESLint, similar to: https://github.com/callstack/eslint-config-callstack#typescript

File: packages/tools/src/isPackagerRunning.ts [hunk 7]
```diff
@@ -16,18 +16,17 @@ import fetch from 'node-fetch';
...
-          body === 'packager-status:running' ? 'running' : 'unrecognized',
-        ),
-    () => 'not_running',
-  );
+  try {
+    const result = await fetch(`http://localhost:${packagerPort}/status`);
+    const body = await result.text();
+
+    return body === 'packager-status:running' ? 'running' : 'unrecognized';
+  } catch (e) {
```
Comment (thymikee) [inline packages/tools/src/isPackagerRunning.ts]:
```suggestion
  } catch (_error) {
```

File: .eslintrc.js (hunk 6 above)
Comment (Esemesek) [reply .eslintrc.js]: TS plugins are included in react-native eslint configuration. https://github.com/facebook/react-native/blob/master/packages/eslint-config-react-native-community/index.js#L48

**Comment by termux-io** [comment]
https://patch-diff.githubusercontent.com/raw/vadimdemedes/ink/pull/9.patch

**Comment by termux-io** [comment]
https://patch-diff.githubusercontent.com/raw/react-native-community/cli/pull/296.patch

File: .vscode/settings.json [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  "prettier.eslintIntegration": true,
+  "eslint.validate": [
+    "javascript",
+    "javascriptreact",
+    {
+      "language": "typescript",
+      "autoFix": true
+    },
+    {
+      "language": "typescriptreact",
```
Comment (grabbou) [inline .vscode/settings.json]: Should we commit `vscode`? I've never included it and kept the local file

File: package.json [hunk 9]
```diff
@@ -4,14 +4,16 @@
...
+    "prebuild": "yarn build:ts",
     "build": "node ./scripts/build.js",
-    "build-clean": "rm -rf ./packages/*/build",
+    "build:ts": "node ./scripts/buildTs.js",
+    "build-clean": "rm -rf ./packages/*/build ./packages/*/tsconfig.tsbuildinfo",
     "watch": "node ./scripts/watch.js",
     "test": "jest",
     "test:ci:unit": "jest packages --ci --coverage",
     "test:ci:e2e": "jest e2e --ci -i",
+    "lint": "eslint --ext .js,.ts . --cache --report-unused-disable-directives",
```
Comment (grabbou) [inline package.json]: Do we need to specify extensions manually even if we use ESLint config?

File: .vscode/settings.json (hunk 8 above)
Comment (thymikee) [reply .vscode/settings.json]: I've added it on purpose because of eslint integration issues that we fixed by adding these entries

File: package.json (hunk 9 above)
Comment (Esemesek) [reply package.json]: For Typescript, yes.

File: packages/tools/src/groupFilesByType.ts [hunk 10]
```diff
@@ -32,5 +21,5 @@ mime.define({
  * the returned object will be: {font: ['fonts/a.ttf'], image: ['images/b.jpg']}
  */
 export default function groupFilesByType(assets: Array<string>) {
-  return groupBy(assets, type => mime.lookup(type).split('/')[0]);
+  return groupBy(assets, type => (mime.getType(type) || '').split('/')[0]);
```
Comment (grabbou) [inline packages/tools/src/groupFilesByType.ts]: When this can return `undefined`?

File: .vscode/settings.json (hunk 8 above)
Comment (Esemesek) [reply .vscode/settings.json]: We should. It gives contributors recommendations for plugins, and provides better integration with vscode eslint plugin. We have to configure additional rule for plugin to check typescript files.

File: packages/tools/src/isPackagerRunning.ts [hunk 11]
```diff
@@ -16,18 +16,17 @@ import fetch from 'node-fetch';
...
-    () => 'not_running',
-  );
+  try {
+    const result = await fetch(`http://localhost:${packagerPort}/status`);
+    const body = await result.text();
+
+    return body === 'packager-status:running' ? 'running' : 'unrecognized';
+  } catch (_error) {
+    return 'not_running';
+  }
```
Comment (grabbou) [inline packages/tools/src/isPackagerRunning.ts]: Feels like a good opportunity for enum here

File: packages/tools/src/groupFilesByType.ts (hunk 10 above)
Comment (Esemesek) [reply packages/tools/src/groupFilesByType.ts]: When you try ot get mime type which doesn't exist.

Comment (grabbou) [reply packages/tools/src/groupFilesByType.ts]: Was it always like that, even with the previous version? Trying to get an idea whether we had a hidden `cannot read property split of undefined` error or the method changed.

Comment (Esemesek) [reply packages/tools/src/groupFilesByType.ts]: Probably yes, because `mime` lib didn't have any flow typings. When upgraded `mime` lib and attaching typescript types, we caught this error.

File: packages/tools/src/isPackagerRunning.ts (hunk 11 above)
Comment (Esemesek) [reply packages/tools/src/isPackagerRunning.ts]: Yes, but the point of this PR was to configure Typescript ecosystem and prepare for migration, not refactoring every LOC.

Comment (grabbou) [reply packages/tools/src/isPackagerRunning.ts]: Well, the PR says "migrate", not "prepare for migration", so that's the reason I am asking. I'd love us to migrate to Typescript fully and embrace its features where possible. 

Shall we open a follow-up issue to track the backlog of things like this, and possibly others that you may already have in mind?

Comment (Esemesek) [reply packages/tools/src/isPackagerRunning.ts]: Ye, sure.

**Comment by termux-io** [comment]
https://weedmaps.com/order-details/confirmation/1947167
